### DISCUSSION TITLE
What are the practical applications of sorting geometries by Hilbert distance

### DISCUSSION THREAD

**Discussion by CNFeffery** [geopandas/geopandas#2893] [Q&A] [answered]
For example, in geopandas `0.13.0`, there are a lot of points that are very close together, but the Hilbert curve values are very different:

```Python
import geopandas as gpd
from shapely.geometry import Point

demo_gdf = gpd.GeoDataFrame(geometry=[Point(x, y) for x in range(100) for y in range(100)])

demo_gdf.sort_values('geometry', inplace=True, ignore_index=True)
demo_gdf['value'] = demo_gdf.index

ax = demo_gdf.plot(column='value', cmap='Reds', s=3);
```

![0adbcba04d324980ec41c76b727eded](https://user-images.githubusercontent.com/49147660/236670048-88749dea-bb51-4f92-bf93-f12cfa137fb2.png)

**Comment by martinfleis** [comment, marked as answer]
Hey, great question!

The goal of the Hilbert curve, or any other space-filling curve is not to ensure that points that are close have always similar distance, that is not possible due to reduction of dimensionality from 2 to 1. The goal is to ensure, that the points that have a similar Hilbert distance are also close in space.

There are several use cases of this, from spatial indexing (see https://en.wikipedia.org/wiki/Hilbert_R-tree) to spatial partitioning. For example, we use it in dask-geopandas to define spatially coherent partitions that are efficient to process in parallel. Using the example above, you can imagine that each of those visible squares would end up being a partition, compared to randomly sorted geometries. See [this documentation page](https://dask-geopandas.readthedocs.io/en/stable/guide/spatial-partitioning.html) for more details.

**Reply by CNFeffery** [reply]
> Hey, great question!
> 
> The goal of the Hilbert curve, or any other space-filling curve is not to ensure that points that are close have always similar distance, that is not possible due to reduction of dimensionality from 2 to 1. The goal is to ensure, that the points that have a similar Hilbert distance are also close in space.
> 
> There are several use cases of this, from spatial indexing (see https://en.wikipedia.org/wiki/Hilbert_R-tree) to spatial partitioning. For example, we use it in dask-geopandas to define spatially coherent partitions that are efficient to process in parallel. Using the example above, you can imagine that each of those visible squares would end up being a partition, compared to randomly sorted geometries. See [this documentation page](https://dask-geopandas.readthedocs.io/en/stable/guide/spatial-partitioning.html) for more details.

I got it, it means that we might be able to effectively group geometries using Hilbert curve values, thus, the acceleration of some gis operations can be realized.

**Reply by martinfleis** [reply]
Precisely. Though it is a bit advanced feature and you don't need it for a vast majority of use cases.
